dashboard/app/controllers/learning_goal_teacher_evaluations_controller.rb

```diff
@@ -26,10 +26,28 @@ def get_evaluation
   end
 
   def get_or_create_evaluation
+    learning_goal = LearningGoal.find(learning_goal_teacher_evaluation_params[:learning_goal_id])
+    @rubric = Rubric.find(learning_goal.rubric_id)
+
+    user_storage_id = storage_id_for_user_id(learning_goal_teacher_evaluation_params[:user_id])
+    script_level = @rubric.lesson.script_levels.find {|sl| sl.levels.include?(@rubric.level)}
+    channel_token = ChannelToken.find_channel_token(
+      script_level.level,
+      user_storage_id,
+      script_level.script_id
+    )
+    raise "channel token not found for user id #{learning_goal_teacher_evaluation_params[:user_id]} and script level id #{script_level.id}" unless channel_token
+    _owner_id, project_id = storage_decrypt_channel_id(channel_token.channel)
+    source_data = SourceBucket.new.get(channel_token.channel, "main.json")
```

Comment:
I wonder whether it's worth it to do this for every update? On one hand, it is possible for a student to unsubmit and resubmit a level before the teacher has submitted feedback. But that feels pretty unlikely so we might be able to only set these fields once.

Thoughts? This may also be something we want to do in the future but is fine as-is for the pilot

Reply:
Good point. I may have to rethink how this function works in order to change that.

Reply:
most of your code should work, just needs to be moved around a bit.

you can likely do 
```
learning_goal_teacher_evaluation = LearningGoalTeacherEvaluation.find_or_create_by!(
      user_id: learning_goal_teacher_evaluation_params[:user_id],
      teacher_id: current_user.id,
      learning_goal_id: learning_goal_teacher_evaluation_params[:learning_goal_id],
)

unless learning_goal_teacher_evaluation.project_id && learning_goal_teacher_evaluation.version_id
  # the code to get those fields
  learning_goal_teacher_evaluation.update!( project_id: project_id, project_version: version_id)
end

```

(might also be nice to only write to the DB once)

I think this does depend on whether we want to optimize for the situation when a student resubmits while a teacher is writing feedback. If we definitely want to have data integrity in that case, I think we will have to query for the project id each time.

Lastly, I think I might've gotten this mixed up with another controller action. This isn't the action that autosave calls right? I'm less concerned if this is only called a handful of times per page load (instead of every time save is called)

Reply:
This is the action that is called when the rubric first loads.

Reply:
It should only be called once

Reply:
once per learning goal right? so a few times on page load but that's it? (this is pendantry now -- I feel a lot better about this implementation now)

Reply:
Yep! That's exactly right. I feel like it's necessary to run the version id checks on each load, because that should prevent a teacher from working on a rubric attached to a previous version of the student's work.